isthmus/src/main/java/io/substrait/isthmus/SqlExpressionToSubstrait.java

```diff
@@ -61,36 +62,66 @@ private record Result(
    */
   public ExtendedExpression convert(String sqlExpression, List<String> createStatements)
       throws SqlParseException {
+    return convert(sqlExpression, ",", createStatements);
+  }
+
+  /**
+   * Converts the given SQL expression string to an {@link io.substrait.proto.ExtendedExpression }
+   *
+   * @param sqlExpression a SQL expression
+   * @param separator the separator for the sql expressions
+   * @param createStatements table creation statements defining fields referenced by the expression
+   * @return a {@link io.substrait.proto.ExtendedExpression }
+   * @throws SqlParseException
+   */
+  public ExtendedExpression convert(
+      String sqlExpression, String separator, List<String> createStatements)
+      throws SqlParseException {
+    return convert(Arrays.asList(sqlExpression.split(separator)), createStatements);
+  }
```

Comment:
In my opinion we should not expose this capability as string splitting like this is very fragile. We should force users to handle their input such that they give us a list of expressions.

Reply:
In case of handling a list of expressions, String[] should be used.

Reply:
As an example, take the following expressions: `'l_orderkey + 9888486986' 'l_orderkey * 2' 'l_orderkey > 10' 'l_orderkey in (10, 20)'`

**Option 1:** You can iterate over the array to get the Extended Expression result. As a result, you will receive 04 extended expressions with 01 expression inside.

**Option 2:** It is also possible that you pass the array of expressions. As a result, you will receive 01 extended expression with 04 expressions inside.

The choice depends on the preferences of the user.

Reply:
Using option 2 here we see all 04 expressions appearing inside `"referredExpr"` as opposed to if you call it separately, it will appear one by one in different responses.

<img width="267" alt="image" src="https://github.com/substrait-io/substrait-java/assets/4554485/5d7d065a-224c-493f-8313-df1d90263221">


````
$ ./isthmus/build/graal/isthmus -c "CREATE TABLE NATION (N_NATIONKEY BIGINT NOT NULL, N_NAME CHAR(25), N_REGIONKEY BIGINT NOT NULL, N_COMMENT VARCHAR(152))" \
    -e  "N_REGIONKEY + 10" "N_REGIONKEY * 2" "N_REGIONKEY > 10" "N_REGIONKEY IN (10, 20)"

{
  "extensionUris": [{
    "extensionUriAnchor": 3,
    "uri": "/functions_boolean.yaml"
  }, {
    "extensionUriAnchor": 1,
    "uri": "/functions_arithmetic.yaml"
  }, {
    "extensionUriAnchor": 2,
    "uri": "/functions_comparison.yaml"
  }],
  "extensions": [{
    "extensionFunction": {
      "extensionUriReference": 1,
      "functionAnchor": 0,
      "name": "add:i64_i64"
    }
  }, {
    "extensionFunction": {
      "extensionUriReference": 1,
      "functionAnchor": 1,
      "name": "multiply:i64_i64"
    }
  }, {
    "extensionFunction": {
      "extensionUriReference": 2,
      "functionAnchor": 2,
      "name": "gt:any_any"
    }
  }, {
    "extensionFunction": {
      "extensionUriReference": 3,
      "functionAnchor": 3,
      "name": "or:bool"
    }
  }, {
    "extensionFunction": {
      "extensionUriReference": 2,
      "functionAnchor": 4,
      "name": "equal:any_any"
    }
  }],
  "referredExpr": [{
    "expression": {
      "scalarFunction": {
        "functionReference": 0,
        "args": [],
        "outputType": {
          "i64": {
            "typeVariationReference": 0,
            "nullability": "NULLABILITY_REQUIRED"
          }
        },
        "arguments": [{
          "value": {
            "selection": {
              "directReference": {
                "structField": {
                  "field": 2
                }
              },
              "rootReference": {
              }
            }
          }
        }, {
          "value": {
            "cast": {
              "type": {
                "i64": {
                  "typeVariationReference": 0,
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "input": {
                "literal": {
                  "i32": 10,
                  "nullable": false,
                  "typeVariationReference": 0
                }
              },
              "failureBehavior": "FAILURE_BEHAVIOR_UNSPECIFIED"
            }
          }
        }],
        "options": []
      }
    },
    "outputNames": ["column-1"]
  }, {
    "expression": {
      "scalarFunction": {
        "functionReference": 1,
        "args": [],
        "outputType": {
          "i64": {
            "typeVariationReference": 0,
            "nullability": "NULLABILITY_REQUIRED"
          }
        },
        "arguments": [{
          "value": {
            "selection": {
              "directReference": {
                "structField": {
                  "field": 2
                }
              },
              "rootReference": {
              }
            }
          }
        }, {
          "value": {
            "cast": {
              "type": {
                "i64": {
                  "typeVariationReference": 0,
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "input": {
                "literal": {
                  "i32": 2,
                  "nullable": false,
                  "typeVariationReference": 0
                }
              },
              "failureBehavior": "FAILURE_BEHAVIOR_UNSPECIFIED"
            }
          }
        }],
        "options": []
      }
    },
    "outputNames": ["column-2"]
  }, {
    "expression": {
      "scalarFunction": {
        "functionReference": 2,
        "args": [],
        "outputType": {
          "bool": {
            "typeVariationReference": 0,
            "nullability": "NULLABILITY_REQUIRED"
          }
        },
        "arguments": [{
          "value": {
            "selection": {
              "directReference": {
                "structField": {
                  "field": 2
                }
              },
              "rootReference": {
              }
            }
          }
        }, {
          "value": {
            "cast": {
              "type": {
                "i64": {
                  "typeVariationReference": 0,
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "input": {
                "literal": {
                  "i32": 10,
                  "nullable": false,
                  "typeVariationReference": 0
                }
              },
              "failureBehavior": "FAILURE_BEHAVIOR_UNSPECIFIED"
            }
          }
        }],
        "options": []
      }
    },
    "outputNames": ["column-3"]
  }, {
    "expression": {
      "scalarFunction": {
        "functionReference": 3,
        "args": [],
        "outputType": {
          "bool": {
            "typeVariationReference": 0,
            "nullability": "NULLABILITY_REQUIRED"
          }
        },
        "arguments": [{
          "value": {
            "scalarFunction": {
              "functionReference": 4,
              "args": [],
              "outputType": {
                "bool": {
                  "typeVariationReference": 0,
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "arguments": [{
                "value": {
                  "selection": {
                    "directReference": {
                      "structField": {
                        "field": 2
                      }
                    },
                    "rootReference": {
                    }
                  }
                }
              }, {
                "value": {
                  "literal": {
                    "i64": "10",
                    "nullable": false,
                    "typeVariationReference": 0
                  }
                }
              }],
              "options": []
            }
          }
        }, {
          "value": {
            "scalarFunction": {
              "functionReference": 4,
              "args": [],
              "outputType": {
                "bool": {
                  "typeVariationReference": 0,
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "arguments": [{
                "value": {
                  "selection": {
                    "directReference": {
                      "structField": {
                        "field": 2
                      }
                    },
                    "rootReference": {
                    }
                  }
                }
              }, {
                "value": {
                  "literal": {
                    "i64": "20",
                    "nullable": false,
                    "typeVariationReference": 0
                  }
                }
              }],
              "options": []
            }
          }
        }],
        "options": []
      }
    },
    "outputNames": ["column-4"]
  }],
  "baseSchema": {
    "names": ["N_NATIONKEY", "N_NAME", "N_REGIONKEY", "N_COMMENT"],
    "struct": {
      "types": [{
        "i64": {
          "typeVariationReference": 0,
          "nullability": "NULLABILITY_REQUIRED"
        }
      }, {
        "fixedChar": {
          "length": 25,
          "typeVariationReference": 0,
          "nullability": "NULLABILITY_NULLABLE"
        }
      }, {
        "i64": {
          "typeVariationReference": 0,
          "nullability": "NULLABILITY_REQUIRED"
        }
      }, {
        "varchar": {
          "length": 152,
          "typeVariationReference": 0,
          "nullability": "NULLABILITY_NULLABLE"
        }
      }],
      "typeVariationReference": 0,
      "nullability": "NULLABILITY_REQUIRED"
    }
  },
  "expectedTypeUrls": []
}
````

Reply:
Fair enough! Thanks for clarifying this @davisusanibar